lib/graphs/abstract_digraph.nit

```diff
@@ -0,0 +1,647 @@
+	# A stack used for Tarjan's algorithm
+	private var tarjan_stack: Queue[V] is noinit
+	# A map associating with each vertex its index
+	private var tarjan_vertex_to_index: HashMap[V, Int] is noinit
+	# A map associating with each vertex its ancestor in Tarjan's algorithm
+	private var tarjan_ancestor: HashMap[V, Int] is noinit
+	# True if and only if the vertex is in the stack
+	private var tarjan_in_stack: HashMap[V, Bool] is noinit
+	# The recursive part of Tarjan's algorithm
+	private fun tarjan(u: V)
+	do
+		tarjan_vertex_to_index[u] = tarjan_index
+		tarjan_ancestor[u] = tarjan_index
+		tarjan_index += 1
+		tarjan_stack.add(u)
+		tarjan_in_stack[u] = true
+		for v in successors(u) do
+			if not tarjan_vertex_to_index.keys.has(v) then
+				tarjan(v)
+				tarjan_ancestor[u] = tarjan_ancestor[u].min(tarjan_ancestor[v])
+			else if tarjan_in_stack[v] then
+				tarjan_ancestor[u] = tarjan_ancestor[u].min(tarjan_vertex_to_index[v])
+			end
+		end
+		if tarjan_vertex_to_index[u] == tarjan_ancestor[u] then
+			var v
+			loop
+				v = tarjan_stack.take
+				tarjan_in_stack[v] = false
+				sccs.union(u, v)
+				if u == v then break
+			end
+		end
+	end
+end
+
+# Mutable digraph
+abstract class MutableDigraph[V, A]
+	super AbstractDigraph[V, A]
+
+	## ---------------- ##
+	## Abstract methods ##
+	## ---------------- ##
+
+	# Adds the vertex `u` to this graph.
+	#
+	# If `u` already belongs to the graph, then nothing happens.
+	#
+	# ~~~
+	# import digraph
+	# var g = new HashMapDigraph[Int, nullable Int]
+	# g.add_vertex(0)
+	# assert g.has_vertex(0)
+	# assert not g.has_vertex(1)
+	# g.add_vertex(1)
+	# assert g.num_vertices == 2
+	# ~~~
+	fun add_vertex(u: V) is abstract
+
+	# Removes the vertex `u` from this graph and all its incident arcs.
+	#
+	# If the vertex does not exist in the graph, then nothing happens.
+	#
+	# ~~~
+	# import digraph
+	# var g = new HashMapDigraph[Int, nullable Int]
+	# g.add_vertex(0)
+	# g.add_vertex(1)
+	# assert g.has_vertex(0)
+	# g.remove_vertex(0)
+	# assert not g.has_vertex(0)
+	# ~~~
+	fun remove_vertex(u: V) is abstract
+
+	# Adds the arc `(u,v)` to this graph.
+	#
+	# If there is an arc from `u` to `v` in this graph, then nothing happens
+	# (even if the value is different).
+	# If vertex `u` or vertex `v` do not exist in the graph, they are added.
+	#
+	# ~~~
+	# import digraph
+	# var g = new HashMapDigraph[Int, nullable Int]
+	# g.add_arc(0, 1)
+	# g.add_arc(1, 2)
+	# assert g.has_arc(0, 1)
+	# assert g.has_arc(1, 2)
+	# assert not g.has_arc(1, 0)
+	# ~~~
+	fun add_arc(u, v: V, l: nullable A) is abstract
+
+	# Changes the value of the arc `(u,v)`.
+	#
+	# If the arc does not exist in the graph, it is created.
+	#
+	# ~~~
+	# import digraph
+	# var g = new HashMapDigraph[Int, nullable Int]
+	# g.add_arc(0, 1, 3)
+	# assert g.get_arc_value(0, 1) == 3
+	# g.change_arc_value(0, 1, 4)
+	# assert g.get_arc_value(0, 1) == 4
+	# ~~~
+	fun change_arc_value(u, v: V, l: A) is abstract
```

Comment:
Hum, what about Nit's generalized setters?

They could allow the user to write:

~~~nit
graph.arc_value(x, y) = 10
print graph.arc_value(x, y) # => 10
~~~

Just rename `get_arc_value` to `arc_value` (in nit, getters do not require the `get` prefix) and rename `change_arc_value` to `arc_value=` (the `=` belong to the name). eg:

~~~
class G
   fun arc_value(u, v: V): A is abstract
   fun arc_value=(u, v: V, l: A) is abstract
end
~~~

The advantage of generalized setter is that it simplify the vocabulary (only one identifier to remember). Another advantage is magic combined assignment, free of charge.

~~~nit
graph.arc_value(x, y) += 1
# equivalent to
graph.arc_value(x, y) = graph.arc_value(x, y) + 1
~~~